api/src/main/java/jakarta/servlet/ServletInputStream.java

```diff
@@ -50,6 +52,69 @@ public abstract class ServletInputStream extends InputStream {
     protected ServletInputStream() {
     }
 
+    /**
+     * Reads from the input stream into the given buffer.
+     * <p>
+     * If the input stream is in non-blocking mode, before each invocation of this method {@link isReady()} must be called
+     * and must return {@code true} or the {@link ReadListener#onDataAvailable()} call back must indicate that data is
```

Comment:
I always interpreted the spec as always requiring `isReady()` to return `true` before you can call `read`, even in an `onDataAvailible` callback. Seeing this made me re-read the relevant sections. The `isReady` section says:

```
The `isReady` method returns `true` if data can be read without blocking. If no data can be
read without blocking it returns `false`. If `isReady` returns `false` it
is illegal to call the read method and an `IllegalStateException` MUST
be thrown.
```

To me that implies that once `isReady` returns false `read` must throw an ISE until it has been called again and returns `true`, even if `onDataAvailable` has been called.

IMHO it makes sense to allow for `read` to be called in `onDataAvailable` before the `isReady` call, but as written I don't think the spec allows it. 

If we are going to allow it we should clarify this on the `isReady` docs as well.

Reply:
Wow, nothing is ever simple!!

So how about this for the input `isReady()` javadoc:

> The `isReady` method returns `true` if it is allowable to call the `read` method. In blocking mode, `isReady` will always return `true`, but a subsequent call to `read` may block awaiting data. In non-blocking mode the `isReady` method may return `false`, in which case it is illegal to call the read method and an `IllegalStateException` MUST be thrown. When `onDataAvailable` is called, a call to `isReady` that returned `true` is implicit.

and for the out `isReady()` javadoc:

The `isReady` method returns `true` if it is allowable to call any method that may write data (e.g. `write`, `print` or `flush`).  In blocking mode, `isReady` will always return `true`, but a subsequent call to method that write data may block. In non-blocking mod the `isReady` method may return false, in which case it is illegal to call a method that writes data and an `IllegalStateException` MUST be thrown. When `onWriteAvailable` is called, a call to `isReady` that returned `true` is implicit.

Reply:
For output, barring fixing a couple of typos, that looks good.
For input, you changed the language from "a read method" to "the `read` method" which begs the question which `read` method? Any of the `read` methods are OK for use in non-blocking mode. The other methods that read data are not. If we changed the text to "a `read` method" that could be sufficient but do we want to consider making calls to the other methods that read data throw an ISE in non-blocking mode or leave the behaviour as undefined?

Reply:
If you are talking about methods like `readAllBytes` think we should mandate an ISE. This is the sort of thing that will work fine in your tests where everything was buffered with your small amount of test data, then fail in production.

Reply:
I'm fine with "a read method" rather than "the read method".

The only doubt I have, for both read/write is if `close` is a method that reads/writes.    I think on input close is not a read. On output close can be a write.. but I also think we treat it specially and allow it to be called at anytime (even though that makes the implementation a lot more complex if there is a pending write in progress).

Reply:
I think it can be called anytime. If we were going to specify it I would say something like:

`If close is called when the stream is in async mode it will immediately return even if the stream contains buffered data that is yet to be written to client. The stream will write this data out in the background, and then close the response. If this process fails the onError method will be invoked as normal.`